src/devices/src/virtio/balloon/utils.rs

```diff
@@ -0,0 +1,190 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::io;
+
+use super::{RemoveRegionError, MAX_PAGES_IN_DESC};
+use vm_memory::{GuestAddress, GuestMemory, GuestMemoryMmap, GuestMemoryRegion};
+
+/// This takes a vector of page frame numbers, and compacts them
+/// into ranges of consecutive pages. The result is a vector
+/// of (start_page_frame_number, range_length) pairs.
+pub(crate) fn compact_page_frame_numbers(v: &mut Vec<u32>) -> Vec<(u32, u32)> {
+    if v.is_empty() {
+        return vec![];
+    }
+
+    // Since the total number of pages that can be
+    // received at once from a single descriptor is `MAX_PAGES_IN_DESC`,
+    // this sort does not change the complexity of handling
+    // an inflation.
+    v.sort();
+
+    // Since there are at most `MAX_PAGES_IN_DESC` pages, setting the
+    // capacity of `result` to this makes sense.
+    let mut result = Vec::with_capacity(MAX_PAGES_IN_DESC);
+
+    // The most recent range of pages is [previous..previous + length).
+    let mut previous = v[0];
+    let mut length = 1;
+
+    for page_frame_number in &v[1..] {
+        // Check if the current page frame number is adjacent to the most recent page range.
+        if *page_frame_number == previous + length {
+            // If so, extend that range.
+            length += 1;
+        } else {
+            // Otherwise, push (previous, length) to the result vector.
+            result.push((previous, length));
+            // And update the most recent range of pages.
+            previous = *page_frame_number;
+            length = 1;
+        }
+    }
+
+    // Don't forget to push the last range to the result.
+    result.push((previous, length));
+
+    result
+}
+
+pub(crate) fn remove_range(
```

Comment:
@alexandruag @serban300 would it make sense to implement (and expose) this in `vm_memory` instead?